### ISSUE TITLE
Add caseOf/either to EitherAsync

### ISSUE THREAD

**Issue by casey-chow** [gigobyte/purify#381]
I notice that one item missing from `EitherAsync` that's in `Either` is the `caseOf` and `either` operators. Right now, it seems like the most fluent way to handle this is to call `.run()` and then fold within the promise:

```ts
await eitherAsync
  .run()
  .then((either) => either.caseOf({
    Left: (error) => `Failure: ${error}`,
    Right: (result) => `Success: ${result}`,
  });
```

But this is fairly awkward in my opinion, especially in contrast with the existing `orDefault`/`leftOrDefault` operators. I think it would make a lot of sense, as a result, for `EitherAsync` to take on the `caseOf` and `either` operators:

```
caseOf<T>(patterns: { Left: (l: L) => T, Right: (r: R) => T } | { _: () => T }): Promise<T>;
either<T>(ifLeft: (value: L) => T, ifRight: (value: R) => T): Promise<T>;
```

**Comment by pyrho** [comment]
You can use [bimap](https://gigobyte.github.io/purify/adts/EitherAsync/#bimap). See [here](https://github.com/gigobyte/purify/issues/304) for another discussion on this topic.

That being said I think it would be beneficial to "synchronise" the APIs of `Either` and `EitherAsync`.

**Comment by gigobyte** [comment]
> That being said I think it would be beneficial to "synchronise" the APIs of Either and EitherAsync.

That's probably the most persuasive point, I also think they should share the same methods and since this has already been requested in the past, I'll tag this as a feature request.

**Comment by gigobyte** [comment]
Added, it's going to get released in `v1.0.0` soon.
